Sources/TuistKit/Services/RunService.swift

```diff
@@ -107,6 +107,13 @@ final class RunService {
 
         try targetRunner.assertCanRunTarget(graphTarget.target)
 
+        let derivedDataPath = try? System.shared.env["CI_DERIVED_DATA_PATH"].map {
```

Comment:
This means we'd respect the `CI_DERIVED_DATA_PATH` only for `tuist run`. Let's move this lower down to the [XcodeProjectBuildDirectoryLocator](https://github.com/tuist/tuist/blob/main/Sources/TuistCore/Utils/XcodeProjectBuildDirectoryLocator.swift#L43-L45).

Also note, `System.shared` shouldn't be used directly as we're moving away from that pattern, so we can enable Swift strict concurrency checking. Instead, follow the same DI pattern as for example [here](https://github.com/tuist/tuist/blob/3599c41d300a5ccf2c877dfe79800c9ca1d3b18e/Sources/TuistSupport/Utils/GitHandler.swift#L48).

You should be able to mock the `CI_DERIVED_DATA_PATH` value in the `MockSystem` and test the functionality in the [XcodeProjectBuildDirectoryLocatorTests](https://github.com/tuist/tuist/blob/3599c41d300a5ccf2c877dfe79800c9ca1d3b18e/Tests/TuistCoreTests/Utils/XcodeProjectBuildDirectoryLocatorTests.swift)

Reply:
@fortmarek Something about this solution didn't sit right with me, so when I got some time today I did a little more digging.

Scanning through my Xcode Cloud log, I found this:

<img width="600" alt="Screenshot 2024-07-08 at 12 17 46 PM" src="https://github.com/tuist/tuist/assets/470753/7af09acd-411d-446c-a4b5-170bb6a4d76e">

I think Xcode Cloud is using the `IDEDerivedDataPathOverride` user default for configuring the derived data dir. This appears to be reproducible locally without Xcode Cloud.

To configure it:

`defaults write com.apple.dt.Xcode IDEDerivedDataPathOverride /Volumes/workspace/DerivedData`

To reset it:

`defaults delete com.apple.dt.Xcode IDEDerivedDataPathOverride`

Also, I think it might be better for Tuist to use something like this to query `xcodebuild` itself for the derived data dir:

`xcodebuild -workspace MyWorkspace.xcworkspace -scheme MyScheme -showBuildSettings | grep -m 1 "CONFIGURATION_BUILD_DIR" | grep -oEi "\/.*"`

In addition to the changes you requested, would you like me to change this PR to query `xcodebuild` for the derived data dir, or read the `IDEDerivedDataPathOverride` user default directly instead of using `CI_DERIVED_DATA_PATH`?

Reply:
Ahh, those are really useful findings!

> In addition to the changes you requested, would you like me to change this PR to query xcodebuild for the derived data dir, or read the IDEDerivedDataPathOverride user default directly instead of using CI_DERIVED_DATA_PATH?

Yes, we should default to the value from that command instead of the env variable 🙌